src/core/probe/kernel/kernel.rs

```diff
@@ -62,20 +56,30 @@ impl fmt::Display for KernelProbe {
     }
 }
 
-struct KernelUnmarshaler {
+#[derive(Default, Deserialize, Serialize, EventSection, EventSectionFactory)]
+pub(crate) struct KernelEvent {
+    /// Kernel symbol name associated with the event (i.e. which probe generated
+    /// the event).
+    pub(crate) symbol: String,
+    /// Probe type: one of "kprobe", "kretprobe" or "raw_tracepoint".
+    pub(crate) probe_type: String,
+    pub(crate) stack_trace: Option<Vec<String>>,
     #[cfg(not(test))]
-    stack_map: libbpf_rs::Map,
+    #[serde(skip)]
+    pub(crate) stack_map: Option<libbpf_rs::Map>,
```

Comment:
Having looked briefly at script pos-processing I'm pretty sure we will end up having to split the `EventSection` and  `EventSectionFactory` objects. Even more if they contain non-clonable objects such as libbpf_rs::Maps.

Do you think we should start by splitting them so we set the "preferred way" of doing things from the very beginning?

Reply:
I have been playing with this and have another alternative:

It's true that associated types cannot be boxed.... but we can can create a binding trait to circumvent that:

Summary:
```rust
pub(crate) trait EventSectionFactory: RawEventSectionFactory, SerdeEventSectionFactory {} // None of these must have associated types

// As promised: no associated types
pub(crate) trait SerdeEventSectionFactory {
    fn from_json(&self, val: serde_json::Value) -> Result<Box<dyn EventSection>>;
}

// This trait does have associated types!
pub(crate) trait EventSectionBinding {
    // The associated section type.
    type Event: for<'a> serde::Deserialize<'a> + EventSection + 'static;

   // A default implementation of the from_json function which uses the fact that Event implements Deserialize.
    fn from_json(&self, val: serde_json::Value) -> Result<Box<dyn EventSection>> {
        Ok(Box::new(serde_json::from_value::<Self::Event>(val)?))
    }
}

// Implement SerdeEventSectionFactory for every type that implements EventSectionBinding
impl<A, B: EventSectionBinding<Event = A>> SerdeEventSectionFactory for B {
    fn from_json(&self, val: serde_json::Value) -> Result<Box<dyn EventSection>> {
        // here Self is treated as  EventSectionBinding
        self.from_json(val)
    }
}
```

Full implementation: https://github.com/amorenoz/retis/commit/39b747ace740d76111378f48844af751141fb6b7

WDYT?

Reply:
Nice!